Fix/prometheusexporter Shutdown HTTP server

Argannor opened the pull request:
**Description:** Shutdown the http.Server instance on exporter shutdown

**Link to tracking Issue:** #35464 

**Testing:** Manual testing. I included this version of the exporter in our internal distribution and deployed it to verify the metrics no longer go stale after receiving a `NOHUP` signal.

**Documentation:** n/a

Argannor committed:
[exporter/prometheus] close server on shutdown

Argannor committed:
[chore] update changelog

dashpole commented on exporter/prometheusexporter/prometheus.go:
```diff
@@ -63,11 +63,14 @@ func (pe *prometheusExporter) Start(ctx context.Context, host component.Host) er
 	srv, err := pe.config.ToServer(ctx, host, pe.settings, mux)
+	pe.shutdownFunc = func(ctx context.Context) error {
+		err := ln.Close()
+		err2 := srv.Shutdown(ctx)
```
will srv be nil if ToServer returned an error above?

Argannor replied:
You're right, I'll adjust it accordingly!

Argannor replied:
I added the nil check in the shutdown handler and kept it where it currently is. Alternatively, we could reassign the shutdownFunc different implementations, but personally I find that more confusing.

dashpole replied:
this looks good to me

atoulme commented on exporter/prometheusexporter/prometheus.go:
```diff
@@ -88,6 +91,6 @@ func (pe *prometheusExporter) ConsumeMetrics(_ context.Context, md pmetric.Metri
-func (pe *prometheusExporter) Shutdown(context.Context) error {
-	return pe.shutdownFunc()
+func (pe *prometheusExporter) Shutdown(ctx context.Context) error {
+	return pe.shutdownFunc(ctx)
```
nil check required

Argannor replied:
Thanks for the hint. What would you prefer to happen when ctx is nil? Pass ctx.TODO on or return an error?

dashpole replied:
Context should never be nil: https://pkg.go.dev/context#pkg-overview.

> Do not pass a nil [Context](https://pkg.go.dev/context#Context), even if a function permits it.

We also don't appear to check for nil context in any other components: https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-collector-contrib+%22ctx+%3D%3D+nil%22&type=code

I would prefer keeping this as-is @atoulme

Argannor committed:
[fix] nil check before server shutdown call

Argannor commented:
Hey @Aneurysm9, @songy23, sorry for pinging you directly. Do you mind having a look at this or advising me what is needed from me to go ahead with this PR?

dashpole committed:
Merge branch 'main' into fix/prometheusexporter-shutdown-server

dashpole commented on .chloggen/fix_prometheusexporter-shutdown-server.yml:
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
```
it looks like the presubmit expects this file to be named with a `.yaml` suffix instead of a `.yml` suffix.  Please update

Argannor committed:
[fix] adjust file suffix to .yaml

dashpole commented on exporter/prometheusexporter/prometheus.go:
```diff
@@ -63,11 +63,17 @@ func (pe *prometheusExporter) Start(ctx context.Context, host component.Host) er
 	mux.Handle("/metrics", pe.handler)
 	srv, err := pe.config.ToServer(ctx, host, pe.settings, mux)
+	pe.shutdownFunc = func(ctx context.Context) error {
+		err := ln.Close()
```
looks like this is failing the linter check since err shadows the err variable outside of the function.  consider renaming this err

Argannor replied:
Sorry about that, renamed both errors, as err2 also wasn't the prettiest of names

Argannor committed:
[refactor] rename err to prevent shadowing

dashpole commented:
unit tests seem to fail after the change.

Argannor commented:
Yeah, I just took a look at the teste that are failing, and I'm not sure what's happening there:

```
	t.Cleanup(func() {
		require.NoError(t, exp.Shutdown(context.Background()))
		// trigger a get so that the server cleans up our keepalive socket
		var resp *http.Response
		resp, err = httpClient.Get("https://localhost:7777/metrics")
		require.NoError(t, err)
		require.NoError(t, resp.Body.Close())
	})
```

Is this a workaround for the fact that the server did not shutdown correctly previously, and is now failing, because after the shutdown the server is shutdown?

dashpole commented:
It might have been.  Try removing the httpClient.Get call from cleanup and see if tests pass.

Argannor committed:
[test] remove cleanup statements

These statements seem to be a workaround for the server not being closed correctly, with the previous changes this should no longer be necessary

Argannor commented:
After removing the statements the tests passed locally for me. Sorry for not running them in the first place, I had some troubles running them using make... :/

Argannor commented:
Is there something you need me to do?
